spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/inbound/ReactiveMongoDbMessageSource.java

```diff
@@ -0,0 +1,243 @@
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.mongodb.inbound;
+
+import java.util.List;
+
+import org.springframework.data.mongodb.MongoDbFactory;
+import org.springframework.data.mongodb.ReactiveMongoDatabaseFactory;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.ReactiveMongoOperations;
+import org.springframework.data.mongodb.core.ReactiveMongoTemplate;
+import org.springframework.data.mongodb.core.convert.MongoConverter;
+import org.springframework.data.mongodb.core.query.BasicQuery;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.expression.Expression;
+import org.springframework.expression.TypeLocator;
+import org.springframework.expression.common.LiteralExpression;
+import org.springframework.expression.spel.support.StandardEvaluationContext;
+import org.springframework.expression.spel.support.StandardTypeLocator;
+import org.springframework.integration.endpoint.AbstractMessageSource;
+import org.springframework.integration.expression.ExpressionUtils;
+import org.springframework.integration.mongodb.support.MongoHeaders;
+import org.springframework.integration.support.AbstractIntegrationMessageBuilder;
+import org.springframework.integration.transaction.IntegrationResourceHolder;
+import org.springframework.transaction.support.TransactionSynchronizationManager;
+import org.springframework.util.Assert;
+
+import com.mongodb.DBObject;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+/**
+ * An instance of {@link org.springframework.integration.core.MessageSource} which returns
+ * a {@link org.springframework.messaging.Message} with a payload which is the result of
+ * execution of a {@link Query}. When expectSingleResult is false (default), the MongoDb
+ * {@link Query} is executed using {@link ReactiveMongoOperations#find(Query, Class)} method which
+ * returns a {@link List}. The returned {@link List} will be used as the payoad of the
+ * {@link org.springframework.messaging.Message} returned by the {{@link #receive()}
+ * method. An empty {@link List} is treated as null, thus resulting in no
+ * {@link org.springframework.messaging.Message} returned by the {{@link #receive()}
+ * method.
+ * <p>
+ * When expectSingleResult is true, the {@link ReactiveMongoOperations#findOne(Query, Class)} is
+ * used instead, and the message payload will be the single object returned from the
+ * query.
+ *
+ * @author David Turanski
+ *
+ * @since 5.2.2
+ */
+public class ReactiveMongoDbMessageSource extends AbstractMessageSource<Object> {
```

Comment:
I think this one should go away for now. 
We need to think more how reactive sources should process a reactive types they get from underlying drivers...